migration/src/main/java/kin/sdk/migration/MigrationManager.java

```diff
@@ -240,21 +280,47 @@ public void run() {
 	}
 
 	private void fireOnReady(final IMigrationManagerCallbacks migrationManagerCallbacks, final IKinClient kinClient,
-		final boolean needToSave) {
+		final String publicAddress, final boolean needToSave) {
 		handler.post(new Runnable() {
 			@Override
 			public void run() {
-				if (needToSave && kinClient.hasAccount()) {
-					IKinAccount account = kinClient.getAccount(kinClient.getAccountCount() - 1);
-					String publicAddress = account.getPublicAddress();
-					saveMigrationCompleted(publicAddress);
-				}
 				isMigrationInProcess.set(false);
-				if (migrationManagerCallbacks != null) {
-					migrationManagerCallbacks.onReady(kinClient);
+				boolean publicAddressIsEmpty = TextUtils.isEmpty(publicAddress);
+				// If no accounts and no public address then that mean we can supply a kinClient object to begin with.
+				// Or if the account has been found in the list of account then we can supply that kinClient which includes this account.
+				if (!kinClient.hasAccount() && publicAddressIsEmpty
+					|| isAccountFoundInListOfAccounts(kinClient, publicAddress)) {
+					if (needToSave && !publicAddressIsEmpty) {
+						saveMigrationCompleted(publicAddress);
+					}
+					if (migrationManagerCallbacks != null) {
+						migrationManagerCallbacks.onReady(kinClient);
+					}
+					// If the account wasn't found in the list of account then we throw the next exception
+				} else {
+					fireOnError(migrationManagerCallbacks, new AccountNotFoundInListOfAccounts());
 				}
 			}
 		});
 	}
 
+	private boolean isAccountFoundInListOfAccounts(IKinClient kinClient, String publicAddress) {
```

Comment:
No unit test for this logic? :(

Reply:
Although we have a different task to add more unit and integration tests i will add a unit tests for this in this branch